components/spinner.py

```diff
@@ -3,29 +3,128 @@
+        self.green = green
+        self.blue = blue
+
+    def dist(self, c) -> tuple:
+        # finds the "distance" of one colour from another
+        # the "distance" is the sum of the differences in rgb value
+        return (
+            abs(self.red - c.red) + abs(self.green - c.green) + abs(self.blue - c.blue)
+        )
+
+
 class Spinner:
     spinner_motor: wpilib.Spark
-    spinner_solenoid: wpilib.DoubleSolenoid
+    spinner_solenoid: wpilib.Solenoid
     colour_sensor: rev.color.ColorSensorV3
 
+    MAX_COLOUR_DIST = 0.2  # Chosen without experimentation, should be tuned
+    COLOUR_OFFSET = 1  # The offset of the colour sensor, in 45 degree increments
+
+    WHEEL_COLOURS = {  # Colours read in real world
+        "R": Colour(82 / 255, 116 / 255, 55 / 255),
+        "B": Colour(55 / 255, 120 / 255, 80 / 255),
+        "G": Colour(61 / 255, 126 / 255, 66 / 255),
+        "Y": Colour(72 / 255, 131 / 255, 50 / 255),
+    }
+
+    WHEEL_ORDER = ["R", "Y", "B", "G"]
+
+    POSITION_SPIN_FACTOR = 0.3  # position control will run at up to twice this speed
+
+    # rotation control uses the formula: 1 - ROTATION_SPIN_FACTOR ** (rotations - ROTATION_TARGET_ROTATIONS)  for wheel speed
+    ROTATION_SPIN_FACTOR = 20  # increase this to recude slow down
+    ROTATION_MAX_SPEED = 1
+    ROTATION_TARGET_ROTATIONS = 3.5
+
     def __init__(self):
         pass
 
     def setup(self):
-        pass
+        self.spinner_motor.set(0)
+        self.raise_wheel()
+        self.rotations = 0
+        # self.state = None
+        self.piston_state = "down"
+        self.required_colour = "B"
+        self.lastCol = "Y"
 
-    def on_enable(self):
-        self.spinner_motor.stopMotor()
-        self.spinner_solenoid.set(wpilib.DoubleSolenoid.Value.kReverse)
+    def on_enable(self) -> None:
+        self.spinner_motor.set(0)
+        self.raise_wheel()
 
-    def execute(self):
-        pass
+    def raise_wheel(self) -> None:  # moves wheel up
+        self.spinner_solenoid.set(False)
+        self.piston_state = "up"
 
-    def read_colour(self):
-        return (self.colour_sensor.getColor(), self.colour_sensor.getIR())
+    def lower_wheel(self) -> None:  # moves wheel down
+        self.spinner_solenoid.set(True)
+        self.piston_state = "down"
 
-    def piston_up(self):
-        self.spinner_solenoid.set(wpilib.DoubleSolenoid.Value.kForward)
+    def get_current_colour(self) -> str:  # reads and catagorizes colour
+        self.sensed_colour = self.colour_sensor.getColor()
 
-    def piston_down(self):
-        self.spinner_solenoid.set(wpilib.DoubleSolenoid.Value.kReverse)
+        distances = {}
+        for col_name, col_value in self.WHEEL_COLOURS.items():
+            distances[col_name] = col_value.dist(self.sensed_colour)
+
+        if (
+            distances[min(distances, key=lambda col: distances[col])]
+            < self.MAX_COLOUR_DIST
+        ):
+            return min(distances, key=lambda col: distances[col])
```

Comment:
This would be much more readable if you only did this once outside this if statement.

Reply:
Also you remove a redundant calculation